### PR TITLE
add twisted support

### PR DESCRIPTION
**Pull request by thrau**
## Motivation

In https://github.com/localstack/localstack/pull/9834 we added support for serving a rolo `Gateway` through twisted, with the limitation that it couldn't support websockets.

This PR moves the basic serving code specific to rolo here, and also adds websocket support for twisted using wsproto and based on changes made in #7. All the code lives in `rolo.serving.twisted`, where imo also `rolo.asgi` should be moved to, to make sure all server-specific code goes into `rolo.serving`.

I did some restructuring of the original code from localstack, for example I added a `TwistedGateway` which can be served via `reactor.listenTCP(4566, TwistedGateway(gateway))`

The test harness still needs a bit of restructuring to make the server parameterization cleaner, currently it's a bit hacky.

After releasing this, I'll raise a PR in localstack to clean up the code there and replace it with the rolo code.

## Changes

* testing framework now adds fixtures for serving things through twisted (with deferred imports though, so dependency is not needed)
* rolo now has constructs for serving a `Gateway` through twisted
* rolo now supports websockets when served through twisted
* several key tests are now run against twisted as well

### CODE REVIEW

File: rolo/serving/twisted.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+    isLeaf = True
+
+    def __init__(
+        self,
+        original: WSGIResource,
+        websocketListener: WebSocketListener,
+    ):
+        super().__init__(original)
+        self.websocketListener = websocketListener
+        self.channel = None
```
Comment (bentsku) [inline rolo/serving/twisted.py:231]: q: should we keep a reference to a channel here? Wouldn't this be overwritten if we get 2 concurrent websockets connections? or is this class instantiated for each incoming request?

**Review by bentsku — COMMENTED** [review]
This is awesome! 🚀 

I've tried to look in details 😄  it took some time to try to understand exactly what was happening and how you supported WebSockets with twisted.

So I'l try to summarize exactly how you've done it, just to verify I've understood right 😄

Twisted does not support websockets out of the box, but you've subclassed the `Channel` class which is the default way of Twisted to implement the "handling" of connection.
This channel is implemented with a queue and a `wsproto` Connection. You manually manipulate the twisted thread pool to be able to have a long lived connection using twisted internals for receiving/sending with the channel, using `websocketListener` to return a `WebSocketListener`, which will under the hood use the `TwistedWebSocketAdapter` with the channel. 

There's one question I'm not too sure I got right:
`WebsocketResourceDecorator`, is this class instantiated for each request? It seems like it is instantiated as part as the resource, which will manage incoming requests. But we hold a reference to the channel created when receiving a `websocket` request in `render`. Wouldn't that be an issue? Maybe we could pass the `channel` around  `_processWebsocket` and `toWsgiEnvironment` instead of having it as an attribute? 
If this class is instantiated for each request then disregard and consider the PR approved. I would have tested my questions locally but I don't know why, I can't run them. 

I had an issue with the `test_handlers` hanging, and when ctrl+c, I would get this exception:
```python
  /home/runner/work/rolo/rolo/rolo/testing/pytest.py:115: RuntimeWarning: coroutine 'BaseEventLoop.shutdown_default_executor' was never awaited
    event_loop.shutdown_default_executor()
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
``` 
Which is also logged in the CI run in this PR. 
The websockets tests are also hanging, but with no error message, so maybe it's something on my end, but I can't test it to answer my questions 😕

File: rolo/serving/twisted.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        request.transport.write(data)
+
+    def wsRespond(
+        self,
+        statusCode: int,
+        extraHeaders: Headers,
+        body: t.Iterator[bytes] | None = None,
+    ):
+        # we could also use self.wsSend(events.RejectConnection(statusCode, ...)) and write manually to the
+        # transport, but instead we re-use twisted's request/response mechanism.
```
[the remark this answers is not in the document]
Comment (thrau) [reply rolo/serving/twisted.py:309]: renamed it to reject and added better docs!

File: rolo/serving/twisted.py (hunk 1 above)
Comment (thrau) [reply rolo/serving/twisted.py:231]: really nice catch! obviously this was nonsensical. cleaned it up to not hold the channel at all.

**Comment by thrau** [comment]
i think i fixed the `SelectReactor` reactor startup, with a big comment inline. thanks for helping me debug this

File: rolo/serving/twisted.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+        if isinstance(request.channel.transport, ProtocolWrapper):
+            request.transport.wrappedProtocol = channel
+        else:
+            request.transport.protocol = channel
+
+        channel.initiateUpgrade()
+
+        environment = to_websocket_environment(request)
+        environment["rolo.websocket"] = TwistedWebSocketAdapter(channel)
+        # WSGIResource also dispatches requests through the threadpool
```
Comment (bentsku) [inline rolo/serving/twisted.py:256]: thanks for the comment 😉

File: rolo/testing/pytest.py [hunk 4]
```diff
@@ -173,7 +174,88 @@ def _serve(gateway: Gateway) -> Server:
...
+    from twisted.internet import reactor
+    from twisted.internet.error import ReactorAlreadyRunning
+    from twisted.web.http import HTTPFactory
+
+    def _run():
+        if reactor.running:
+            return
+
+        try:
+            # for some reason, when using a `SelectReactor` (like you do by default on MacOS), whatever
```
Comment (bentsku) [inline rolo/testing/pytest.py:189]: thank you for the fix 🙏

**Review by bentsku — APPROVED** [review]
LGTM! Thanks a lot for fixing the tests and for all the additional info we went over. This is great looking, also thanks for the `accept/reject` logic, it seems a bit clearer now. 

Awesome 🚀 can't wait for it to be default at some point 😉

**Pull request merged**